chore: Improve main logging

uristdwarf opened the pull request:
Improve logging to better debug errors and where they occured, also use safs functions.

Closes #221

[Only merge after #234]

uristdwarf committed:
chore: Improve main logging

Improve logging to better debug errors and where they occured, also use
safs functions.

Closes #221

Signed-off-by: Urmas Rist <urmas@leil.io>

lgsilva3087 commented on src/main/main.cc:
```diff
@@ -983,6 +986,7 @@ int main(int argc,char **argv) {
 	// Only kStart should check for lock file consistency
 	FileLock fl(runmode, locktimeout);
 	if (fl.lockstatus() == FileLock::LockStatus::kFail) {
+		safs::log_err("couldn't not acquire lock on {}", fl.name());
```
Maybe 'not' is not needed in the sentence.

lgsilva3087 reviewed (dismissed):
LGTM. Please see my comment.

uristdwarf committed:
fix(main): Logs breaking -p behavior

Due to (for whatever reason on earth) main logging the output of the pid
into stderr, any script expecting only this PID in stderr will break.
This fixes it by exiting early after the PID has been acquired

Signed-off-by: Urmas Rist <urmas@leil.io>

uristdwarf committed:
chore(main): Fix typo "couldn't not" -> "couldn't"

Signed-off-by: Urmas Rist <urmas@leil.io>

wg1k commented on src/main/main.cc:
```diff
@@ -151,7 +151,7 @@ bool initialize(const std::vector<RunTab> &tabs) {
 			}
 		} catch (const std::exception &e) {
-			safs_pretty_syslog(LOG_ERR, "%s", e.what());
+			safs::log_err("unhandled exception ({}): {}", typeid(e).name(), e.what());
```
consider the usage of the recently created `log_error_code`:
```suggestion
			safs::log_error_code(e, "unhandled exception");
```

uristdwarf replied:
Doesn't work with exceptions. I've added a log_exception to slogger based on your for errno

wg1k commented on src/main/main.cc:
```diff
@@ -656,18 +660,19 @@ void makedaemon() {
 	fflush(stderr);
 	if (pipe(piped)<0) {
-		safs_pretty_syslog(LOG_ERR, "pipe error");
+		safs::log_err("pipe error: {}", strerr(errno));
```
In case you want to use it, `log_error_code` also handles `errno`:
```suggestion
		safs::log_error_code(errno, "pipe error");
```

wg1k commented on src/main/main.cc:
```diff
@@ -656,18 +660,19 @@ void makedaemon() {
 	f = fork();
 	if (f<0) {
-		safs_pretty_errlog(LOG_ERR, "first fork error");
+		safs::log_err("pipe error: {}", strerr(errno));
```
```suggestion
		safs::log_error_code(errno, "pipe error");
```

wg1k commented on src/main/main.cc:
```diff
@@ -656,18 +660,19 @@ void makedaemon() {
 		wait(&f);       // just get child status - prevents child from being zombie during initialization stage
 		if (f) {
 			safs_pretty_syslog(LOG_ERR, "child status: %d",f);
+			safs::log_err("{}", strerr(errno));
```
Is it possible to unify both error messages?

uristdwarf replied:
Actually, now that I look at it, the log_err should check if wait returned -1

wg1k commented on src/main/main.cc:
```diff
@@ -678,12 +683,13 @@ void makedaemon() {
 						happy = fwrite(pipebuff,1,r-1,stderr);
 						(void)happy;
 					}
+					safs::log_err("pipe error: {}", strerr(errno));
```
```suggestion
					safs::log_error_code(errno, "pipe error");
```

wg1k commented on src/main/main.cc:
```diff
@@ -678,12 +683,13 @@ void makedaemon() {
 				(void)happy;
 			} else {
-				safs_pretty_errlog(LOG_ERR,"error reading pipe");
+				safs::log_err("pipe error: {}", strerr(errno));
```
```suggestion
				safs::log_error_code(errno, "pipe error");
```

wg1k commented on src/main/main.cc:
```diff
@@ -693,9 +699,9 @@ void makedaemon() {
+		safs::log_err("second fork error: {}", strerr(errno));
 		if (write(piped[1],"fork error\n",11)!=11) {
-			safs_pretty_errlog(LOG_ERR,"pipe write error");
+			safs::log_err("pipe error: {}", strerr(errno));
```
```suggestion
			safs::log_error_code(errno, "pipe error");
```

wg1k commented on src/main/main.cc:
```diff
@@ -693,9 +699,9 @@ void makedaemon() {
 	f = fork();
 	if (f<0) {
-		safs_pretty_errlog(LOG_ERR,"second fork error");
+		safs::log_err("second fork error: {}", strerr(errno));
```
```suggestion
		safs::log_error_code(errno, "second fork error");
```

wg1k commented on src/main/main.cc:
```diff
@@ -948,7 +954,7 @@ int main(int argc,char **argv) {
 
 	if (chdir(wrkdir)<0) {
-		safs_pretty_syslog(LOG_ERR,"can't set working directory to %s",wrkdir);
+		safs::log_err("can't set working directory to {}: {}", wrkdir, strerr(errno));
```
```suggestion
		safs::log_error_code(errno, "can't set working directory to {}", wrkdir);s
```

uristdwarf committed:
chore: Use new safs::log_error_code

* Add safs::log_exception for exception logging

Signed-off-by: Urmas Rist <urmas@leil.io>

lgsilva3087 commented on src/main/main.cc:
```diff
@@ -533,8 +534,11 @@ FileLock::LockStatus FileLock::wdlock(RunMode runmode, uint32_t timeout) {
 			return LockStatus::kAlive;
 		}
 		if (runmode==RunMode::kTest) {
+			// TODO(5.0.0): Fix this stupidity in a breaking change by changing it
+			// to stdout. It's not that hard, stderr for logs/errors, stdout
+			// for output
```
Not nice comment.

uristdwarf replied:
Trust me, my initial comment was much meaner :) I lost a lot of hours because of this.
It won't last long, I plan to change this to stdout very soon.

lgsilva3087 commented on src/slogger/slogger.h:
```diff
@@ -138,6 +138,20 @@ inline void log_error_code(int error_code, const FormatType &format, Args &&...a
+ * @param format The format string for the log message.
+ * @param args The arguments to format the log message.
+ */
+template <typename FormatType, typename... Args>
+inline void log_exception(const std::exception &exception, const FormatType &format, Args &&...args) {
+    auto extended_format = fmt::format("{} (exception: {}, message: {})", format,
+                                     typeid(exception).name(), exception.what());
+    safs::log_err(extended_format, std::forward<Args>(args)...);
+}
+
```
Please apply the our format for new code.

lgsilva3087 reviewed (approved):
LGTM
